Adding documentation for lxplus (#2756)
ECP-WarpX/WarpX: Docs/source/install/hpc/lxplus.rst

.. _building-lxplus:

LXPLUS (CERN)
=============

The LXPLUS cluster is located at CERN.

* `Lxplus documentation <https://lxplusdoc.web.cern.ch>`__
* Batch system: `HTCondor <https://batchdocs.web.cern.ch/index.html>`__
* Filesystem locations:
    * User folder: ``/afs/cern.ch/user/<a>/<account>`` (10GByte)
    * Work folder: ``/afs/cern.ch/work/<a>/<account>`` (100GByte)
    * Eos storage: ``/eos/home-<a>/<account>`` (1T)

> Through LXPLUS we have access to CPU and GPU nodes (the latter equipped with NVIDIA A100 or T4 GPUs).

Reviewer: I find it useful to write how much memory each GPU has, because it makes it easier to translate runs between different machines. Do you like to add this info for the two GPU types?

I usually add this info in a `Running` section, which might still be missing here (example job scripts). See, e.g.,:
- https://warpx.readthedocs.io/en/latest/install/hpc/summit.html#running
- https://warpx.readthedocs.io/en/latest/install/hpc/cori.html#running

Is there anything interesting to add or link for a `Post-Processing` section as well? :)
- https://warpx.readthedocs.io/en/latest/install/hpc/summit.html#post-processing
- https://warpx.readthedocs.io/en/latest/install/hpc/cori.html#post-processing

Installation
------------
Only very little software is pre-installed on Lxplus so we show how to install from scratch all the dependencies using `Spack <https://spack.io>`__.

For size reasons it is not advisable to install WarpX in the ``$HOME`` directory, while it should be installed in the "work directory". For this purpose we set an environment variable with the path to the "work directory"

.. code-block:: bash

    export WORK=/afs/cern.ch/work/<a>/<account>/

We clone WarpX in ``$WORK``:

.. code-block:: bash

    cd $WORK

>     git clone https://github.com/ECP-WarpX/WarpX.git

Reviewer: Everyone loves lowercase directories for faster `cd` typing:
```suggestion
    git clone https://github.com/ECP-WarpX/WarpX.git warpx
```

.. code-block:: bash

    spack env activate warpx-lxplus

If we are planning on running with GPU support then we must set the environment variable ``SPACK_STACK_USE_CUDA``

.. code-block:: bash

    SPACK_STACK_USE_CUDA=1 spack env create warpx-lxplus-cuda $WORK/WarpX/Tools/machines/lxplus-cern/spack.yaml
    spack env activate warpx-lxplus-cuda

and if we want to use the python interface we must set the environment variable ``SPACK_STACK_USE_PYTHON``.
So if we want both CUDA-acceleration and build a Python interface, the environment setup would look like this:

.. code-block:: bash

    SPACK_STACK_USE_PYTHON=1 SPACK_STACK_USE_CUDA=1 SPACK_STACK_USE_CUDA=1 spack env create warpx-lxplus-cuda-py $WORK/WarpX/Tools/machines/lxplus-cern/spack.yaml
    spack env activate warpx-lxplus-cuda-py

Then we can install the required packages:

.. code-block:: bash

    spack install

The environment ``warpx-lxplus`` (or ``-cuda`` or ``-cuda-py``) must be reactivated everytime that we log in so it could be a good idea to add the following lines to the ``.bashrc``:

.. code-block:: bash

>     cd $WORK

Reviewer: Not needed with a named environment:
```suggestion
```

But we need to activate Spack first:
```suggestion
    source $WORK/spack/share/spack/setup-env.sh
```

.. _building-lxplus:

LXPLUS (CERN)
=============

The LXPLUS cluster is located at CERN.

* `Lxplus documentation <https://lxplusdoc.web.cern.ch>`__
* Batch system: `HTCondor <https://batchdocs.web.cern.ch/index.html>`__
* Filesystem locations:
    * User folder: ``/afs/cern.ch/user/<a>/<account>`` (10GByte)
    * Work folder: ``/afs/cern.ch/work/<a>/<account>`` (100GByte)
    * Eos storage: ``/eos/home-<a>/<account>`` (1T)

Through LXPLUS we have access to CPU and GPU nodes (the latter equipped with NVIDIA V100).

Installation
------------
Only very little software is pre-installed on Lxplus so we show how to install from scratch all the dependencies using `Spack <https://spack.io>`__.

For size reasons it is not advisable to install WarpX in the ``$HOME`` directory, while it should be installed in the "work directory". For this purpose we set an environment variable with the path to the "work directory"

.. code-block:: bash

>     export WORK=/afs/cern.ch/work/<a>/<account>/

Reviewer: As in your profile:
```suggestion
    export WORK=/afs/cern.ch/work/${USER:0:1}/$USER/
```

We clone WarpX in ``$WORK``:

.. code-block:: bash

    cd $WORK
    git clone https://github.com/ECP-WarpX/WarpX.git warpx

> GCC

Reviewer: For all the steps below, do you want to advertise just copying the `Tools/machines/lxplus-cern/lxplus_warpx.profile.example` script into `$HOME` and then `source`-ing it?

Author: Perfect, should that be the only option or we also keep the instructions?

Reviewer: Keeping is fine if you like :)

^^^^^^^^^^^^^^^^^^^^^^^^^
The easiest way to install the dependencies is to use the pre-prepared ``warpx.profile`` as follows:

.. code-block:: bash

>     cd $WORK
>     cp warpx/WarpX/Tools/machines/lxplus-cern/lxplus_warpx.profile.example .
>     source lxplus_warpx.profile.example

Reviewer: Let's remove the `.example` :) We add this so people know that they might need to edit this file (not the case here).
```suggestion
    cp $WORK/warpx/WarpX/Tools/machines/lxplus-cern/lxplus_warpx.profile.example $WORK/lxplus_warpx.profile
    source $WORK/lxplus_warpx.profile
```

Author: Alright, I was actually wondering why it was called example :)

> When doing this one can directly skip to the **Bulding Warpx** section.

Reviewer: Nice little link:
```suggestion
When doing this one can directly skip to the :ref:`Building WarpX <building-lxplus-warpx>` section.
```

To have the environment activated at every login it is then possible to add the following lines to the ``.bashrc``

.. code-block:: bash

    export WORK=/afs/cern.ch/work/${USER:0:1}/$USER/

>     source $WORK/lxplus_warpx.profile.example

Reviewer: ```suggestion
    source $WORK/lxplus_warpx.profile
```

.. code-block:: bash

    spack compiler find /cvmfs/sft.cern.ch/lcg/releases/gcc/11.2.0-ad950/x86_64-centos7/bin

Installing the Dependencies
^^^^^^^^^^^^^^^^^^^^^^^^^^^

To install the dependencies we create a virtual environment, which we call ``warpx-lxplus``:

.. code-block:: bash

    spack env create warpx-lxplus $WORK/WarpX/Tools/machines/lxplus-cern/spack.yaml
    spack env activate warpx-lxplus
    spack install

If the GPU support or the Python bindings are not needed, it's possible to skip the installation by respectively setting
the following environment variables export ``SPACK_STACK_USE_PYTHON=0`` and ``export SPACK_STACK_USE_CUDA = 0`` before
running the previous commands.

After the installation is done once, all we need to do in future sessions is just ``activate`` the environment again:

.. code-block:: bash

    spack env activate warpx-lxplus

The environment ``warpx-lxplus`` (or ``-cuda`` or ``-cuda-py``) must be reactivated everytime that we log in so it could
be a good idea to add the following lines to the ``.bashrc``:

.. code-block:: bash

    source $WORK/spack/share/spack/setup-env.sh
    spack env activate -d warpx-lxplus
    cd $HOME

> Building WarpX

Reviewer: Link anchor/reference:
```suggestion
.. _building-lxplus-warpx:

Building WarpX
```

^^^^^^^^^^^^^^

We prepare and load the Spack software environment as above.
Then we build WarpX:

.. code-block:: bash

    cmake -S . -B build
    cmake --build build -j 6

Or if we need to compile with CUDA:

.. code-block:: bash

>     cmake -S . -B build -DWarpX_COMPUTE=CUDA -DAMReX_CUDA_ARCH='7.0;7.5'

Reviewer: This should not be needed if we set the env variable in our `warpx.profile` :)
```suggestion
    cmake -S . -B build -DWarpX_COMPUTE=CUDA
```

    cmake --build build -j 6

Python Bindings
^^^^^^^^^^^^^^^

Here we assume that a Python interpreter has been set up as explained previously.

> Then we compile WarpX as in the previous section (with or without CUDA) adding ``-DWarpX_LIB=ON`` and then we install it into our Python:
> 
> .. code-block:: bash
> 
>     PYWARPX_LIB_DIR=$PWD/build/lib python3 -m pip wheel .
>     python3 -m pip install pywarpx-*whl

Reviewer: Yes, and since #2822:
```suggestion
Now, ensure Python tooling is up-to-date:
.. code-block:: bash

   python3 -m pip install -U pip setuptools wheel

Then we compile WarpX as in the previous section (with or without CUDA) adding ``-DWarpX_LIB=ON`` and then we install it into our Python:

.. code-block:: bash

   cmake -S . -B build -DWarpX_COMPUTE=CUDA -DWarpX_LIB=ON
   cmake --build build --target pip_install -j 6

This builds WarpX for 3D geometry.

Alternatively, if you like to build WarpX for all geometries at once, use:

.. code-block:: bash

   BUILD_PARALLEL=6 python3 -m pip wheel .
   python3 -m pip install pywarpx-*whl
```

.. _building-lxplus:

LXPLUS (CERN)
=============

The LXPLUS cluster is located at CERN.

* `Lxplus documentation <https://lxplusdoc.web.cern.ch>`__
* Batch system: `HTCondor <https://batchdocs.web.cern.ch/index.html>`__
* Filesystem locations:
    * User folder: ``/afs/cern.ch/user/<a>/<account>`` (10GByte)
    * Work folder: ``/afs/cern.ch/work/<a>/<account>`` (100GByte)
    * Eos storage: ``/eos/home-<a>/<account>`` (1T)

Through LXPLUS we have access to CPU and GPU nodes (the latter equipped with NVIDIA A100 or T4 GPUs).

Installation
------------
Only very little software is pre-installed on Lxplus so we show how to install from scratch all the dependencies using `Spack <https://spack.io>`__.

For size reasons it is not advisable to install WarpX in the ``$HOME`` directory, while it should be installed in the "work directory". For this purpose we set an environment variable with the path to the "work directory"

.. code-block:: bash

    export WORK=/afs/cern.ch/work/<a>/<account>/

We clone WarpX in ``$WORK``:

.. code-block:: bash

    cd $WORK
    git clone https://github.com/ECP-WarpX/WarpX.git

GCC
^^^

> The pre-installed GNU compiler is outdated so we need a more recent compiler. Here we use the gcc 9.2.0 from the LCG project, but other options are possible.

Reviewer: ```suggestion
The pre-installed GNU compiler is outdated so we need a more recent compiler. Here we use the gcc 11.2.0 from the LCG project, but other options are possible.
```

We activate it by doing

.. code-block:: bash

>     source /cvmfs/sft.cern.ch/lcg/releases/gcc/9.2.0/x86_64-centos7/setup.sh

Reviewer: ```suggestion
    source /cvmfs/sft.cern.ch/lcg/releases/gcc/11.2.0-ad950/x86_64-centos7/setup.sh
```

> In order to avoid using different compilers this line could be added directly into the ``$HOME/.bashrc`` file.

Reviewer: To be discussed: we can create as always a `warpx.profile` file that is manually sourced on login when a WarpX workflow is planned.
This file can contain all needed activations (compiler + spack + spack env) and would then not interfere with workflows on lxplus where you need totally different software dependencies.

Spack
^^^^^
We download and activate Spack in ``$WORK``:

.. code-block:: bash

    cd $WORK
    git clone -c feature.manyFiles=true https://github.com/spack/spack.git
    source spack/share/spack/setup-env.sh

When installing packages Spack will try to set permissions in a way which is forbidden on the LXPLUS file system (AFS), but we can avoid this by setting the Spack option:

.. code-block:: bash

    spack config add "config:allow_sgid:false"

> Now we add our gcc 9.2.0 compiler to spack:

Reviewer: Following your latest update:
```suggestion
Now we add our gcc 11.2.0 compiler to spack:
```

.. code-block:: bash

>     spack compiler find /cvmfs/sft.cern.ch/lcg/releases/gcc/9.2.0-afc57/x86_64-centos7/bin/

Reviewer: ```suggestion
    spack compiler find /cvmfs/sft.cern.ch/lcg/releases/gcc/11.2.0-ad950/x86_64-centos7/bin
```

Installing the Dependencies
^^^^^^^^^^^^^^^^^^^^^^^^^^^

> To install the depndencies we create an anonymous environment.
> First of all we copy the envornmnet file in a local folder with the desired name of the environment (e.g. ``warpx-lxplus``):
> 
> .. code-block:: bash
> 
>     cd $WORK
>     mkdir warpx-lxplus
>     cp $WORK/WarpX/Tools/machines/lxplus-cern/spack.yaml $WORK/warpx-lxplus/.
> 
> Then we activate the environment:
> 
> .. code-block:: bash
> 
>     spack env activate -d warpx-lxplus

Reviewer: I just realized in that case, we can just create a _named_ environment :D

```suggestion
To install the dependencies we create a virtual environment, which we call ``warpx-lxplus``:

.. code-block:: bash

    spack env create warpx-lxplus $WORK/WarpX/Tools/machines/lxplus-cern/spack.yaml
    spack env activate warpx-lxplus
    spack install

After the installation is done once, all we need to do in future sessions is just ``activate`` the environment again:

.. code-block:: bash

    spack env activate warpx-lxplus
```

If we are planning on running with GPU support then we must set the environment variable ``SPACK_STACK_USE_CUDA``

.. code-block:: bash

>     export SPACK_STACK_USE_CUDA=1

Reviewer: ```suggestion 
    SPACK_STACK_USE_CUDA=1 spack env create warpx-lxplus-cuda $WORK/WarpX/Tools/machines/lxplus-cern/spack.yaml
    spack env activate warpx-lxplus-cuda
```

and if we want to use the python interface we must set the environment variable ``SPACK_STACK_USE_PYTHON``

> .. code-block:: bash
> 
>     export SPACK_STACK_USE_PYTHON=1

Reviewer: ```suggestion
.. code-block:: bash

    SPACK_STACK_USE_PYTHON=1 SPACK_STACK_USE_CUDA=1 SPACK_STACK_USE_CUDA=1 spack env create warpx-lxplus-cuda-py $WORK/WarpX/Tools/machines/lxplus-cern/spack.yaml
    spack env activate warpx-lxplus-cuda-py
```

> and if we want to use the python interface we must set the environment variable ``SPACK_STACK_USE_PYTHON``

Reviewer: ```suggestion
and if we want to use the python interface we must set the environment variable ``SPACK_STACK_USE_PYTHON``.
So if we want both CUDA-acceleration and build a Python interface, the environment setup would look like this:
```

Then we can install the required packages:

.. code-block:: bash

    spack install

> The environment ``warpx-lxplus`` must be reactivated everytime that we log in so it could be a good idea to add the following lines to the ``.bashrc``:

Reviewer: ```suggestion
The environment ``warpx-lxplus`` (or ``-cuda`` or ``-cuda-py``) must be reactivated everytime that we log in so it could be a good idea to add the following lines to the ``.bashrc``:
```

.. code-block:: bash

    cd $WORK

>     spack env activate -d warpx-lxplus

Reviewer: Note: Spack activation via `source spack/share/spack/setup-env.sh` also needs to go there. But see comment on `warpx.profile` workflow above.

Reviewer: added now :)

    cd $HOME

Building WarpX
^^^^^^^^^^^^^^

We prepare and load the Spack software environment as above.
Then we build WarpX:

.. code-block:: bash

    cmake -S . -B build

>     cmake --build build

Reviewer: If you like, add a sensible parallelism number here, which is kind to the head-node.
E.g.,
```suggestion
    cmake --build build -j 6
```

Or if we need to compile with CUDA:

.. code-block:: bash

    cmake -S . -B build -DWarpX_COMPUTE=CUDA -DAMReX_CUDA_ARCH='7.0;7.5'

>     cmake --build build

Reviewer: If this is a sensible number:
```suggestion
    cmake --build build -j 6
```

.. _building-lxplus:

LXPLUS (CERN)
=============

The LXPLUS cluster is located at CERN.

* `Lxplus documentation <https://lxplusdoc.web.cern.ch>`__
* Batch system: `HTCondor <https://batchdocs.web.cern.ch/index.html>`__
* Filesystem locations:

>     * User folder: `/afs/cern.ch/user/<a>/<account>` (10GByte)
>     * Work folder: `/afs/cern.ch/work/<a>/<account>` (100GByte)
>     * Eos storage: `/eos/home-<a>/<account>` (1T)

Reviewer: Verbatim in `.rst` is like this:
```suggestion
    * User folder: ``/afs/cern.ch/user/<a>/<account>`` (10GByte)
    * Work folder: ``/afs/cern.ch/work/<a>/<account>`` (100GByte)
    * Eos storage: ``/eos/home-<a>/<account>`` (1T)
```

Through LXPLUS we have access to CPU and GPU nodes (the latter equipped with NVIDIA A100 or T4 GPUs).

Installation
------------

> Only very little software is pre-installed on Lxplus so we show how to install from scratch all the dependencies using Spack.

Reviewer: For users jumping right into this section, we can link Spack here:
```suggestion
Only very little software is pre-installed on Lxplus so we show how to install from scratch all the dependencies using `Spack <https://spack.io>`__.
```

> For size reasons it is not advisable to install WarpX in the $HOME directory, while it should be installed in the "work directory". For this purpose we set an environment variable with the path to the "work directory"

Reviewer: ```suggestion
For size reasons it is not advisable to install WarpX in the ``$HOME`` directory, while it should be installed in the "work directory". For this purpose we set an environment variable with the path to the "work directory"
```

.. code-block:: bash

    export WORK=/afs/cern.ch/work/<a>/<account>/

GCC
^^^
The pre-installed GNU compiler is outdated so we need a more recent compiler. Here we use the gcc 9.2.0 from the LCG project, but other options are possible.

We activate it by doing

.. code-block:: bash

    source /cvmfs/sft.cern.ch/lcg/releases/gcc/9.2.0/x86_64-centos7/setup.sh

In order to avoid using different compilers this line could be added directly into the ``$HOME/.bashrc`` file.

Spack
^^^^^
We download and activate Spack in ``$WORK``:

.. code-block:: bash

    cd $WORK
    git clone -c feature.manyFiles=true https://github.com/spack/spack.git
    source spack/share/spack/setup-env.sh

> When installing packages Spack will try to set permissions in a way which is forbidden on the LXPLUS file system (AFS), but we can avoid this by modifying the ``spack/etc/spack/defaults/config.yaml``. Inside this file we need to change the line
> 
> .. code-block:: bash
> 
>     allow_sgid: true
> 
> to
> 
> .. code-block:: bash
> 
>     allow_sgid: false

Reviewer: Easier to run a command than changing a file (and the files in `spack/etc/spack/defaults/` should not be changed usually, but their counter-parts in `$HOME/.spack/` are):
```suggestion
When installing packages Spack will try to set permissions in a way which is forbidden on the LXPLUS file system (AFS), but we can avoid this setting the Spack option:

.. code-block:: bash

    spack config add "config:allow_sgid:false"
```

Author: Oh I didn't know this one. Thanks a lot 😄

Now we add our gcc 9.2.0 compiler to spack:

.. code-block:: bash

    spack compiler find /cvmfs/sft.cern.ch/lcg/releases/gcc/9.2.0-afc57/x86_64-centos7/bin/

Installing the Dependencies
^^^^^^^^^^^^^^^^^^^^^^^^^^^

With the following commands we install the dependencies for WarpX:

.. code-block:: bash

>     spack install ncurses+symlinks %gcc@9.2.0

Reviewer: Individual `spack install` + `spack load` commands can potentially load incompatible software. A better way to install this is using a virtual environment, [like here](https://warpx.readthedocs.io/en/latest/install/dependencies.html#spack-macos-linux).

All packages and the option + compiler above can be summarized in a single "download this `.yaml` environment file", which might be easiest to do here.

Example of such a file: https://github.com/openPMD/openPMD-api/blob/dev/.github/ci/spack-envs/clang8_py38_mpich_h5_ad1_ad2/spack.yaml

- `specs:` section - the packages I want
- `packages:` section - mainly selecting what I want to re-use from existing packages on the system
- `compilers:` section - the system compiler I like to configure & use
- `config:` section - currently only used to limit the -j compilation to two cores

In order to populate the `compilers:` section, you can copy out from `spack config edit compilers` in your current setup.

Using this file will also allow us to make `gcc@9.2.0` the default compiler, so we don't need to type it all the time, by making it the first and only entry in `compilers:`.

For CUDA, you could either do an additional environment file or add an option like is done in #2719 to add the additional CUDA package and variants.

    spack install adios2 ^ncurses+symlinks %gcc@9.2.0
    spack install blaspp ^ncurses+symlinks %gcc@9.2.0
    spack install cmake ^ncurses+symlinks %gcc@9.2.0
    spack install ccache ^ncurses+symlinks %gcc@9.2.0
    spack install fftw ^ncurses+symlinks %gcc@9.2.0
    spack install hdf5 ^ncurses+symlinks %gcc@9.2.0
    spack install lapackpp ^ncurses+symlinks %gcc@9.2.0
    spack install mpi ^ncurses+symlinks %gcc@9.2.0
    spack install openpmd-api ^ncurses+symlinks %gcc@9.2.0
    spack install pkgconfig %gcc@9.2.0

If we are planning on running WarpX on GPUs we also need to install Cuda:

.. code-block:: bash

    spack install cuda ^ncurses+symlinks %gcc@9.2.0

Now we load the installed modules:

.. code-block:: bash

    spack load cmake
    spack load openmpi
    spack load ccache

And if we installed Cuda:

.. code-block:: bash

    spack load cuda

Building WarpX
^^^^^^^^^^^^^^

> We download WarpX in ``$WORK``:

Reviewer: ```suggestion
We prepare and load the Spack software environment as above.
Then we download WarpX in ``$WORK``:
```

.. code-block:: bash

    cd $WORK
    git clone https://github.com/ECP-WarpX/WarpX.git
    cd WarpX

Then we build WarpX:

.. code-block:: bash

    cmake -S . -B build
    cmake --build build

> Or if we need to compile with cuda:

Reviewer: ```suggestion
Or if we need to compile with CUDA:
```

.. code-block:: bash

    cmake -S . -B build -DWarpX_COMPUTE=CUDA -DAMReX_CUDA_ARCH='7.0;7.5'
    cmake --build build

Python Bindings
^^^^^^^^^^^^^^^

> Here we assume that a Python interpreter has been set up (e.g. with Spack or a Miniconda).

Reviewer: ```suggestion
Here we assume that a Python interpreter has been set up (e.g., with Spack or a Miniconda).
```

Reviewer: I would not mix in (Mini)Conda with Spack here, since Conda brings in its own compilers and stdlibs, which likely will cause very hard to debug issues down the road.
The easiest way to build the Python bindings is to add the spack packages `python` and `py-pip`, then do the steps below.

Author: Ok then I'll just add `python` and `py-pip` them to the environment file I think

> Then we compile WarpX as in the previous section (with or without Cuda) adding ``-DWarpX_LIB=ON`` and then we install it into our Python:

Reviewer: ```suggestion
Then we compile WarpX as in the previous section (with or without CUDA) adding ``-DWarpX_LIB=ON`` and then we install it into our Python:
```

------------
Only very little software is pre-installed on Lxplus so we show how to install from scratch all the dependencies using `Spack <https://spack.io>`__.

For size reasons it is not advisable to install WarpX in the ``$HOME`` directory, while it should be installed in the "work directory". For this purpose we set an environment variable with the path to the "work directory"

.. code-block:: bash

    export WORK=/afs/cern.ch/work/<a>/<account>/

We clone WarpX in ``$WORK``:

.. code-block:: bash

    cd $WORK
    git clone https://github.com/ECP-WarpX/WarpX.git

GCC
^^^
The pre-installed GNU compiler is outdated so we need a more recent compiler. Here we use the gcc 11.2.0 from the LCG project, but other options are possible.

We activate it by doing

.. code-block:: bash

    source /cvmfs/sft.cern.ch/lcg/releases/gcc/11.2.0-ad950/x86_64-centos7/setup.sh

In order to avoid using different compilers this line could be added directly into the ``$HOME/.bashrc`` file.

Spack
^^^^^
We download and activate Spack in ``$WORK``:

.. code-block:: bash

    cd $WORK
    git clone -c feature.manyFiles=true https://github.com/spack/spack.git
    source spack/share/spack/setup-env.sh

> When installing packages Spack will try to set permissions in a way which is forbidden on the LXPLUS file system (AFS), but we can avoid this by setting the Spack option:
> 
> .. code-block:: bash
> 
>     spack config add "config:allow_sgid:false"
> 

Reviewer: With out new environment file, we can just pre-configure this and the user does not even need to do this step anymore :)
```suggestion
```

.. code-block:: bash

    cmake -S . -B build -DWarpX_COMPUTE=CUDA
    cmake --build build -j 6

Python Bindings
^^^^^^^^^^^^^^^

Here we assume that a Python interpreter has been set up as explained previously.

Now, ensure Python tooling is up-to-date:

> .. code-block:: bash

Reviewer: Upsi, newline:
```suggestion

.. code-block:: bash
```
